File: sdk/python/src/covenant/enforcer.py
from __future__ import annotations

import time
from datetime import datetime, timezone
from functools import wraps
from pathlib import Path
from typing import Any, Awaitable, Callable

from covenant.audit import AuditEmitter, AuditEvent
from covenant.errors import CovenantViolationError
from covenant.interceptor import ToolCall, ToolCallInterceptor, _active_interceptor
from covenant.invariants import InvariantEvaluator
from covenant.loader import load_spec
from covenant.models import CovenantSpec
from covenant.validator import validate_input, validate_output

_emitter = AuditEmitter()


class ContractEnforcer:
    """Stateless enforcer loaded once per @contract decoration.

    Args:
        spec: Fully-validated CovenantSpec.
        spec_path: Path to the .covenant.yaml (used to resolve protocol schema paths).
    """

    def __init__(self, spec: CovenantSpec, spec_path: Path) -> None:
        self._spec = spec
        self._spec_dir = spec_path.parent

    async def enforce(
        self, fn: Callable[..., Awaitable[Any]], *args: Any, **kwargs: Any
    ) -> Any:
        """Run the full enforcement flow around a single agent invocation.

        Enforcement sequence:
            1. Validate input against protocols.input.schema
            2. Set ContextVar to activate the OpenAI SDK interceptor
            3. Await the wrapped agent function
            4. Validate output against protocols.output.schema
            5. Run invariant assertions
            6. Emit audit event (fire-and-forget)
            7. Return output

        Args:
            fn: The wrapped async agent function.
            *args: Positional arguments forwarded to fn.
            **kwargs: Keyword arguments forwarded to fn.

        Returns:
            Whatever fn returns.

        Raises:
            CovenantViolationError: On any constraint or invariant breach.
        """
        start = time.monotonic()

        # 1. Input validation
        input_data = args[0] if args else kwargs
        input_schema = (
            self._spec.protocols.input.schema_path
            if self._spec.protocols and self._spec.protocols.input
            else None
        )
        validate_input(input_data, input_schema, self._spec_dir)

        # 2. Install per-invocation interceptor via ContextVar
        interceptor = ToolCallInterceptor(self._spec)
        token = _active_interceptor.set(interceptor)

        outcome = "pass"
        violation_code: str | None = None
        output: Any = None

        try:
            output = await fn(*args, **kwargs)
        except CovenantViolationError as exc:
            outcome = "violation"
            violation_code = exc.code
            raise
        finally:
            _active_interceptor.reset(token)

        # 3. Output validation
        output_schema = (
            self._spec.protocols.output.schema_path
            if self._spec.protocols and self._spec.protocols.output
            else None
        )
        validate_output(output, output_schema, self._spec_dir)

        # 4. Invariant evaluation
        duration_ms = (time.monotonic() - start) * 1000
        if self._spec.invariants:
            evaluator = InvariantEvaluator(self._spec.invariants)
            try:
                evaluator.evaluate(
                    {
                        "input": input_data,
                        "output": output,
                        "tool_calls": interceptor.tool_calls,
                        "duration_ms": duration_ms,
                        "cost_usd": interceptor.total_cost_usd,
                        "model_used": interceptor.model_used,
                    }
                )
            except CovenantViolationError as exc:
                outcome = "violation"
                violation_code = exc.code
                raise

        # 5. Emit audit event (fire-and-forget)
        _emitter.emit(
            AuditEvent(
                contract=f"{self._spec.agent.name}@{self._spec.agent.version}",
                outcome=outcome,
                tool_calls=[
                    {"name": tc.name, "args": tc.args}
                    for tc in interceptor.tool_calls
                ],
                cost_usd=interceptor.total_cost_usd,
                duration_ms=duration_ms,
                occurred_at=datetime.now(timezone.utc),
                violation_code=violation_code,
                model_used=interceptor.model_used,
            )
        )

        return output


def contract(spec_path: str | Path) -> Callable:
    """Decorator factory that wraps an async agent function with contract enforcement.

    The spec is loaded once at decoration time (import time). Any spec parse error
    raises immediately — not at call time.

    Args:
        spec_path: Path to the .covenant.yaml file.

    Returns:
        Decorator that wraps an async function with enforcement.

    Example:
        @contract("./my-agent.covenant.yaml")
        async def run(input: dict) -> dict:
            ...
    """
    path = Path(spec_path)
    spec = load_spec(path)
    enforcer = ContractEnforcer(spec, path)

    def decorator(fn: Callable[..., Awaitable[Any]]) -> Callable:
        @wraps(fn)
        async def wrapper(*args: Any, **kwargs: Any) -> Any:
            return await enforcer.enforce(fn, *args, **kwargs)

        return wrapper

    return decorator


async def call_tool(
    name: str, fn: Callable[..., Awaitable[Any]], /, *args: Any, **kwargs: Any
) -> Any:
    """Explicit enforcement layer for non-LLM tool calls.

    Use this to enforce tool constraints for filesystem, HTTP, subprocess, or any
    other I/O that doesn't go through the OpenAI SDK interceptor.

    If called outside a @contract invocation (no active interceptor), the function
    executes directly with no enforcement applied.

    Args:
        name: Tool name to check against capabilities.tools and constraints.deny_tools.
        fn: Async callable to invoke if the tool is permitted.
        *args: Positional arguments forwarded to fn.
        **kwargs: Keyword arguments forwarded to fn.

    Returns:
        Whatever fn returns.

    Raises:
        CovenantViolationError: UNDECLARED_TOOL or DENIED_TOOL if the tool is not allowed.

    Example:
        result = await call_tool("read_file", read_file_async, path="./data.csv")
    """
    interceptor = _active_interceptor.get()
    if interceptor is not None:
        interceptor._check_tool(name)

    result = await fn(*args, **kwargs)

    if interceptor is not None:
        interceptor.tool_calls.append(
            ToolCall(
                name=name,
                args=kwargs,
                result=result,
                cost_usd=0.0,
            )
        )

    return result

File: sdk/typescript/src/enforcer.ts
import { dirname, resolve } from "node:path";
import { AuditEmitter } from "./audit.js";
import { CovenantViolationError } from "./errors.js";
import {
  ToolCallInterceptor,
  interceptorStorage,
  installPatches,
} from "./interceptor.js";
import { evaluateInvariants } from "./invariants.js";
import { loadSpec } from "./loader.js";
import type { CovenantSpec } from "./spec.js";
import { validateInput, validateOutput } from "./validator.js";

// Install OpenAI SDK patches once at module load time
installPatches();

const _emitter = new AuditEmitter();

// ── Types ─────────────────────────────────────────────────────────────────────

type AsyncFn<A extends unknown[], R> = (...args: A) => Promise<R>;

// ── ContractEnforcer ──────────────────────────────────────────────────────────

export class ContractEnforcer {
  constructor(
    private readonly spec: CovenantSpec,
    private readonly specDir: string
  ) {}

  async enforce<A extends unknown[], R>(
    fn: AsyncFn<A, R>,
    args: A
  ): Promise<R> {
    const start = Date.now();
    const inputData = args[0];

    // 1. Validate input
    validateInput(inputData, this.spec.protocols?.input?.schema, this.specDir);

    // 2. Create per-invocation interceptor and run fn inside AsyncLocalStorage
    const interceptor = new ToolCallInterceptor(this.spec);

    let outcome: "pass" | "violation" = "pass";
    let violationCode: string | null = null;
    let output: R;

    try {
      output = await interceptorStorage.run(interceptor, () => fn(...args));
    } catch (err) {
      if (err instanceof CovenantViolationError) {
        outcome = "violation";
        violationCode = err.code;
      }
      throw err;
    }

    // 3. Validate output
    validateOutput(output, this.spec.protocols?.output?.schema, this.specDir);

    // 4. Evaluate invariants
    const durationMs = Date.now() - start;
    if (this.spec.invariants && this.spec.invariants.length > 0) {
      try {
        evaluateInvariants(this.spec.invariants, {
          input: inputData,
          output,
          toolCalls: interceptor.toolCalls,
          durationMs,
          costUsd: interceptor.totalCostUsd,
          modelUsed: interceptor.modelUsed,
        });
      } catch (err) {
        if (err instanceof CovenantViolationError) {
          outcome = "violation";
          violationCode = err.code;
        }
        throw err;
      }
    }

    // 5. Emit audit event (fire-and-forget)
    _emitter.emit({
      contract: `${this.spec.agent.name}@${this.spec.agent.version}`,
      outcome,
      toolCalls: interceptor.toolCalls.map((tc) => ({
        name: tc.name,
        args: tc.args,
      })),
      costUsd: interceptor.totalCostUsd,
      durationMs,
      occurredAt: new Date().toISOString(),
      violationCode,
      modelUsed: interceptor.modelUsed,
    });

    return output;
  }
}

// ── contract() decorator factory ──────────────────────────────────────────────

/**
 * Wraps an async function with Covenant contract enforcement.
 * The spec is loaded synchronously at call time (typically module load).
 *
 * @param specPath - Path to the .covenant.yaml file
 * @returns Decorator that wraps an async function with full enforcement
 *
 * @example
 * const run = contract("./my-agent.covenant.yaml")(
 *   async (input: Input): Promise<Output> => { ... }
 * );
 */
export function contract(
  specPath: string
): <A extends unknown[], R>(fn: AsyncFn<A, R>) => AsyncFn<A, R> {
  const absPath = resolve(specPath);
  const spec = loadSpec(absPath);
  const specDir = dirname(absPath);
  const enforcer = new ContractEnforcer(spec, specDir);

  return function <A extends unknown[], R>(fn: AsyncFn<A, R>): AsyncFn<A, R> {
    return async (...args: A): Promise<R> => {
      return enforcer.enforce(fn, args);
    };
  };
}

// ── callTool() ────────────────────────────────────────────────────────────────

/**
 * Explicit enforcement layer for non-LLM tool calls (filesystem, HTTP, subprocess).
 *
 * Checks the tool name against capabilities.tools and constraints.deny_tools.
 * If called outside a contract() invocation, executes fn directly (no enforcement).
 *
 * @param name - Tool name to check against the spec
 * @param fn - Async function to invoke if the tool is permitted
 * @param args - Arguments forwarded to fn
 *
 * @example
 * const data = await callTool("read_file", readFileAsync, "./data.csv");
 */
export async function callTool<A extends unknown[], R>(
  name: string,
  fn: (...args: A) => Promise<R>,
  ...args: A
): Promise<R> {
  const interceptor = interceptorStorage.getStore();
  if (interceptor !== undefined) {
    interceptor.checkTool(name);
  }

  const result = await fn(...args);

  if (interceptor !== undefined) {
    interceptor.toolCalls.push({
      name,
      args: args[0] !== undefined ? (args[0] as Record<string, unknown>) : {},
      result,
      costUsd: 0,
      timestamp: new Date(),
    });
  }

  return result;
}
